File: Python Implemantation/SlidingWindowCounter.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#
# @author: Uluc Furkan Vardar
# @updatedDate: 10.03.2020
# @version: 1.0.0

try:
	import redis
	from redis.client import Redis
	import time
	import random
except Exception as e:
	print ('Libs import error, ', str(e) )
	exit()

class SlidingWindowCounter():
	def __init__(self, host, port, db, windowInSecond):
		self.windowInSecond = windowInSecond
		self.host = host
		self.port = port
		self.db = db
		self.r = Redis( host = self.host, port=self.port, db=self.db) 

	def increment(self, key):
		currentMs = int(round(time.time() * 1000))		
		maxScoreMs = currentMs - self.windowInSecond*1000
		
		self.r.zremrangebyscore( key, 0, maxScoreMs )
		self.r.zadd( key, {currentMs : currentMs + random.random()} )
		self.r.expire( key, self.windowInSecond)

	def getCount(self, key):
		return self.r.zcard(key)

File: Java Implemantation/src/main/java/slidingWindowCounter/SlidingWindowCounter.java
package slidingWindowCounter;

import redis.clients.jedis.JedisPool;
import redis.clients.jedis.Transaction;

/**
 * Created by halil on 31/10/16.
 */
public class SlidingWindowCounter {

    private static SlidingWindowCounter instance  = null;
    private        JedisPool            jedisPool = null;

    private SlidingWindowCounter() {
    }

    public static SlidingWindowCounter getInstance() {
        if (instance == null) {
            synchronized (SlidingWindowCounter.class) {
                if (instance == null) {
                    instance = new SlidingWindowCounter();
                }
            }
        }

        return instance;
    }

    public void configure(String host, int port) throws Exception {
        jedisPool = new JedisPool(host, port);
    }

    public void configure(JedisPool pool) throws Exception {
        jedisPool = pool;
    }

    public void increment(String key, int windowInSecond) throws Exception {

        long currentMs = System.currentTimeMillis();
        long maxScoreMs = currentMs - windowInSecond * 1000;

        Transaction redis = jedisPool.getResource().multi();
        redis.zremrangeByScore(key, 0, maxScoreMs);
        redis.zadd(key, currentMs, currentMs + "-" + Math.random()); //keep members unique
        redis.expire(key, windowInSecond);
        redis.exec();
    }

    public long getCount(String key) throws Exception {
        return jedisPool.getResource().zcard(key);
    }
}
